sdk/servicebus/service-bus/src/util/utils.ts

```diff
@@ -485,3 +487,96 @@ export type EntityStatus =
   | "Renaming"
   | "Restoring"
   | "Unknown";
+
+/**
+ * An executor for a Promise (or function that returns a Promise) that obeys both a timeout and an
+ * optional AbortSignal.
+ * @param timeoutMs - The number of milliseconds to allow before throwing an AMQP error with condition ServiceUnavailableError
+ * @param timeoutMessage - The message to place in the .description field for the thrown exception for Timeout.
+ * @param abortSignal - The abortSignal associated with containing operation.
+ * @param abortErrorMsg - The abort error message associated with containing operation.
+ * @param value - The value to be resolved with after a timeout of t milliseconds.
+ * @returns {Promise<T>} - Resolved promise
+ *
+ * @internal
+ * @ignore
+ */
+export function waitForTimeoutOrAbortOrResolve<T>(args: {
+  actionFn: () => Promise<T>;
+  timeoutMs: number;
+  timeoutMessage: string;
+  abortSignal?: AbortSignalLike;
+  abortMessage?: string;
+}): Promise<T> {
+  if (args.abortSignal && args.abortSignal.aborted) {
+    throw new AbortError(args.abortMessage);
+  }
+
+  let timer: any | undefined = undefined;
+  let clearAbortSignal: (() => void) | undefined = undefined;
+
+  const clearAbortSignalAndTimer = (): void => {
+    clearTimeout(timer);
+
+    if (clearAbortSignal) {
+      clearAbortSignal();
+    }
+  };
+
+  const abortOrTimeoutPromise = new Promise((resolve, reject) => {
+    clearAbortSignal = checkAndRegisterWithAbortSignal(reject, args.abortMessage, args.abortSignal);
+
+    // using a named function here so we can identify it in our unit tests
+    timer = setTimeout(function timeoutCallback() {
+      reject({
+        condition: ErrorNameConditionMapper.ServiceUnavailableError,
+        description: args.timeoutMessage
+      } as AmqpError);
+    }, args.timeoutMs);
+  });
+
+  const actionPromise = args.actionFn();
+  return Promise.race([abortOrTimeoutPromise, actionPromise])
+    .then(() => {
+      clearAbortSignalAndTimer();
+      return actionPromise;
+    })
+    .catch((err) => {
+      clearAbortSignalAndTimer();
+      throw err;
```

Comment:
(Copied over from https://github.com/Azure/azure-sdk-for-js/pull/8994#discussion_r430900519)

Since `clearListenersAndTimer()` needs to be called in both success and failure case, why not use finally?
```js
try {
     return Promise.race([abortAndTimeoutPromise, actionPromise])
} finally {
     clearListenersAndTimer();
}
```

Reply:
I believe to do that I'd have to await on Promise.race(), right? Otherwise all the clear logic could run before the promises have actually completed?

I've tried to make it so the function does all the prep in a non-async way (including an initial check on the abort signal).

Reply:
> I believe to do that I'd have to await on Promise.race(), right? Otherwise all the clear logic could run before the promises have actually completed?

That is true, you will have to await Promise.race()

> I've tried to make it so the function does all the prep in a non-async way (including an initial check on the abort signal).

`waitForTimeoutOrAbortOrResolve` is returning a promise. Not adding await does not make the function non-async.

Reply:
Not to belabor the point but I wanted to make it clear - the Promise itself is async. The code _before_ it is the part that is async. The function itself is not declared async - this was to address an issue you'd brought up in a prior PR where I'd migrated a check into an async function and I believe you had perf concerns about it.

I chatted with @chradek about this offline and he made a pretty good point that one issue with this style (sync + async throwing) is that the user can't attach just a single type of error handling to the function. If they try/catch everything works fine but if they .catch() on it they won't actually get the error I throw above.  

So I think I can make everyone happy by just marking the entire function async :)

Reply:
This has been done!

Reply:
Can you elaborate the case where try/catch would find the error but .catch() wont? Perhaps a small code snippet?

Reply:
Sure.

So in the old code:

```typescript
function doThis() {
   if (whatever) {
     throw new Error();
  }

  return new Promise();
}

// #1: try/catch
try {
  await doThis();
} catch (err) {
}

// #2: .catch()
await doThis().catch(err);
```

In #1 try/catch will catch everything, sync or errors thrown from the promise that's returned.
In #2 .catch() will only catch errors from the promise that's returned, not from the sync code that runs before it.

Now if the function itself was marked as async then everything works just fine either way (which is what I ended up doing).

Another way that we (@chradek  and I tried) that wouldn't help is just doing Promise.reject(err) instead of throwing inside of the function (when it's not async). That apparently still causes a microtask to be scheduled so you'd still not get "immediate" execution.

Reply:
sweet, thanks!